### PR TITLE
refactor: adapted temp module to satisfy conventions

### LINKED ISSUE #15: task: refactor and test temp module
### Description:

Refactor the temp module of the cellboard and add the missing unit tests.

### Implementation steps:

1. Change all the identifiers name following the embedded conventions
2. Implement unit tests
3. Fix small issues if any

### Dependencies:

None

### Additional notes:

_No response_

### PR DESCRIPTION
**Pull request by ale-giustina**
# Purpose

Refactoring of temp module
# Overview

Removed typedef on structs and enums, added doxygen comments and added unit tests.
# Guidance

The main files that have been affected are temp.h, temp-api.h, temp-api.c and test_volt.c. Other files may have been modified marginally due to variable renaming.

### CODE REVIEW

File: Core/Src/bms/temp-api.c
Comment (Tonidotpy) [inline Core/Src/bms/temp-api.c:1]: File does not comply with coding standards.
Double check macros and identifier names and change them following the appropriate conventions.
Remember to use LibEagleTRT whenever possible.

File: test/test_temp/test-temp.c
Comment (Tonidotpy) [inline test/test_temp/test-temp.c:1]: There should also be tests for the conversion function from voltages to temperature in °C.

File: test/test_temp/test-temp.c [hunk 1]
```diff
@@ -1,166 +1,283 @@
...
+FAKE_VOID_FUNC(mock_set_address, uint8_t);
+FAKE_VOID_FUNC(mock_start_conversion);
+
+void test_temp_init_ok() {
+
+    const enum TempReturnCode rc = temp_init(mock_set_address, mock_start_conversion);
 
-void mock_start_conversion(void) {
+    TEST_ASSERT_EQUAL_MESSAGE(TEMP_OK, rc, "temp_init() failed to return TEMP_OK");
 }
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c:102]: Missing asserts for the handler content that are split in multiple tests.

File: test/test_temp/test-temp.c [hunk 2]
```diff
@@ -1,166 +1,283 @@
...
+    TEST_ASSERT_EQUAL_MESSAGE(TEMP_NULL_POINTER, rc, "temp_init() should fail with NULL start_conversion");
 }
 
-void test_temp_init_cellboard_id() {
-    TEST_ASSERT_EQUAL_MESSAGE(CELLBOARD_ID, htemp.temp_can_payload.cellboard_id, "Cellboard ID mismatch in payload initialization");
+void test_temp_init_correct_cellboard_id() {
+    TEST_ASSERT_EQUAL_MESSAGE(CELLBOARD_ID, temp_handler.temp_can_payload.cellboard_id, "Cellboard ID mismatch in payload initialization");
 }
 
 void test_temp_start_conversion_ok() {
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: The behavior you are testing here is not much clear, at least from the name.

File: test/test_temp/test-temp.c [hunk 3]
```diff
@@ -1,166 +1,283 @@
...
+
+    TEST_ASSERT_EQUAL_MESSAGE(TEMP_OK, rc, "temp_start_conversion() failed to return TEMP_OK");
+    TEST_ASSERT_TRUE_MESSAGE(temp_handler.busy, "Module should be busy after starting conversion");
+    TEST_ASSERT_EQUAL_MESSAGE(1U, mock_set_address_fake.call_count, "temp_start_conversion() should call set_address() once");
+    TEST_ASSERT_EQUAL_MESSAGE(1U, mock_start_conversion_fake.call_count, "temp_start_conversion() should call start_conversion() once");
+    TEST_ASSERT_EQUAL_MESSAGE(1U, mock_set_address_fake.arg0_val, "temp_start_conversion() should select the next mux address");
 }
 
 void test_temp_start_conversion_busy() {
     temp_start_conversion();
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: You should not be calling the same function twice just because it sets the busy flag to true.
The function must be called during the Act phase after the Arrange phase.
If you are testing the behavior when busy is set to true you should change its value and then call the function (i.e. Arrange -> Act).

File: test/test_temp/test-temp.c [hunk 4]
```diff
@@ -1,166 +1,283 @@
...
+
+    TEST_ASSERT_EQUAL_MESSAGE(TEMP_BUSY, rc, "temp_start_conversion() should return TEMP_BUSY if already busy");
 }
 
-void test_temp_notify_conversion_complete() {
-    temp_start_conversion();
+void test_temp_notify_conversion_complete_ok() {
 
-    // Values in Volts to trigger polynomial conversion
+    temp_start_conversion();
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: Avoid calling multiple function, you should arrange what you need to test the notify function and not call the start conversion, otherwise you are not doing unit testing since this is not limited to a "unit" of code.

File: test/test_temp/test-temp.c [hunk 5]
```diff
@@ -1,166 +1,283 @@
...
-    TEST_ASSERT_FLOAT_WITHIN_MESSAGE(0.01f, 20.0f, htemp.temperatures[0], "First value mismatch");
-    TEST_ASSERT_FLOAT_WITHIN_MESSAGE(0.01f, 30.0f, htemp.temperatures[1], "Second value mismatch");
+
+    celsius values[2] = { 20.0f, 30.0f };
+
+    const enum TempReturnCode rc = temp_update_values(3U, values, 2U);
+
+    TEST_ASSERT_EQUAL_MESSAGE(TEMP_OK, rc, "temp_update_values() failed to return TEMP_OK");
+    TEST_ASSERT_FLOAT_WITHIN_MESSAGE(0.01f, 20.0f, temp_handler.temperatures[3], "First value mismatch");
+    TEST_ASSERT_FLOAT_WITHIN_MESSAGE(0.01f, 30.0f, temp_handler.temperatures[4], "Second value mismatch");
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: You should use the `ARRAY` assert variant here.

File: test/test_temp/test-temp.c [hunk 6]
```diff
@@ -1,166 +1,283 @@
...
 void test_temp_update_discharge_value_ok() {
-    // 2.0V corresponds to a specific temp in the polynomial, just checking return code here
-    TEST_ASSERT_EQUAL_MESSAGE(TEMP_OK, temp_update_discharge_value(0, 2.0f), "temp_update_discharge_value() failed");
+
+    temp_handler.discharge_temperatures[3] = 0.0f;
+
+    const enum TempReturnCode rc = temp_update_discharge_value(3U, 2.0f);
+
+    TEST_ASSERT_EQUAL_MESSAGE(TEMP_OK, rc, "temp_update_discharge_value() failed to return TEMP_OK");
+    TEST_ASSERT_NOT_EQUAL_MESSAGE(0.0f, temp_handler.discharge_temperatures[3], "temp_update_discharge_value() failed to update the discharge temperature");
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: Why are you checking that the updated value is not equal to zero? This assertion may be true even if the temperature value is different from the one expected.

File: test/test_temp/test-temp.c [hunk 7]
```diff
@@ -1,166 +1,283 @@
...
+
+    temp_handler.discharge_temperatures[1] = -999.0f;
+    temp_handler.discharge_temperatures[2] = -999.0f;
+    volt_t values[2] = { 2.0f, 2.1f };
+
+    const enum TempReturnCode rc = temp_update_discharge_values(1U, values, 2U);
+
+    TEST_ASSERT_EQUAL_MESSAGE(TEMP_OK, rc, "temp_update_discharge_values() failed to return TEMP_OK");
+    TEST_ASSERT_NOT_EQUAL_MESSAGE(-999.0f, temp_handler.discharge_temperatures[1], "First discharge temperature was not updated");
+    TEST_ASSERT_NOT_EQUAL_MESSAGE(-999.0f, temp_handler.discharge_temperatures[2], "Second discharge temperature was not updated");
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: You should use the `ARRAY` assert variant and compare for equality instead.

File: Core/Src/bms/temp-api.c [hunk 8]
```diff
@@ -88,158 +88,158 @@ _STATIC_INLINE void _temp_check_cells_value(const size_t index, const celsius_t
...
-const cells_temp_t *temp_get_values(void) {
-    return &htemp.temperatures;
+const cells_temp *temp_get_values(void) {
+    return &temp_handler.temperatures;
 }
 
-celsius_t temp_get_min(void) {
-    celsius_t min = TEMP_MAX_C;
+celsius temp_get_min(void) {
+    celsius min = TEMP_MAX_C;
```
Comment (Tonidotpy) [inline Core/Src/bms/temp-api.c]:
```suggestion
    celsius min = temp_handler.temperatures[0];
```

File: Core/Src/bms/temp-api.c [hunk 9]
```diff
@@ -88,158 +88,158 @@ _STATIC_INLINE void _temp_check_cells_value(const size_t index, const celsius_t
...
-        min = CELLBOARD_MIN(min, htemp.temperatures[i]);
+        min = CELLBOARD_MIN(min, temp_handler.temperatures[i]);
     }
     return min;
 }
 
-celsius_t temp_get_max(void) {
-    celsius_t max = 0U;
+celsius temp_get_max(void) {
+    celsius max = 0U;
```
Comment (Tonidotpy) [inline Core/Src/bms/temp-api.c]:
```suggestion
    celsius max = temp_handler.temperatures[0];
```

File: test/test_temp/test-temp.c [hunk 10]
```diff
@@ -1,166 +1,283 @@
...
+    for (size_t i = 0U; i < sizeof(out) / sizeof(out[0]); ++i)
+        expected[i] = temp_handler.temperatures[CELLBOARD_SEGMENT_TEMP_SENSOR_COUNT / 2 + i];
 
-    TEST_ASSERT_EQUAL_MESSAGE(sizeof(htemp.temp_can_payload), size, "Payload size mismatch");
-    TEST_ASSERT_EQUAL_MESSAGE(0, payload->offset, "Payload offset mismatch");
-    TEST_ASSERT_FLOAT_WITHIN_MESSAGE(0.01f, 10.0f, payload->temperature_0, "Payload temp 0 mismatch");
+    const enum TempReturnCode rc = temp_dump_values(out, CELLBOARD_SEGMENT_TEMP_SENSOR_COUNT / 2U, CELLBOARD_SEGMENT_TEMP_SENSOR_COUNT / 2U - 1U);
+
+    TEST_ASSERT_EQUAL_MESSAGE(TEMP_OK, rc, "temp_dump_values() failed to return TEMP_OK for valid input");
+    TEST_ASSERT_EQUAL_MEMORY_MESSAGE(expected, out, sizeof(out), "temp_dump_values() failed to copy the correct values to the output array");
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: Use the `FLOAT_ARRAY` assert variant instead of `MEMORY`.

File: test/test_temp/test-temp.c [hunk 11]
```diff
@@ -1,166 +1,283 @@
...
+void test_temp_dump_values_ok() {
+    for (size_t i = 0U; i < CELLBOARD_SEGMENT_TEMP_SENSOR_COUNT; ++i)
+        temp_handler.temperatures[i] = 10.0f + (celsius)i;
 
-    htemp.offset = 0; // Reset offset manually for deterministic test
+    celsius out[CELLBOARD_SEGMENT_TEMP_SENSOR_COUNT / 2 - 1U];
 
-    bms_cellboard_cells_temperature_converted_t *payload = temp_get_cells_temp_canlib_payload(&size);
+    celsius expected[CELLBOARD_SEGMENT_TEMP_SENSOR_COUNT / 2 - 1U];
+    for (size_t i = 0U; i < sizeof(out) / sizeof(out[0]); ++i)
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: Not really a fan of the expression to get the size of an array using `sizeof`, you may use the macro definition directly or use an auxilitary variable instead.

File: test/test_temp/test-temp.c [hunk 12]
```diff
@@ -1,166 +1,283 @@
...
+    temp_handler.temperatures[2] = 12.0f;
+    temp_handler.temperatures[3] = 13.0f;
+
+    bms_cellboard_cells_temperature_converted_t expected_payload = { 0 };
+    expected_payload.cellboard_id = CELLBOARD_ID;
+    expected_payload.offset = 0U;
+    expected_payload.temperature_0 = 10.0f;
+    expected_payload.temperature_1 = 11.0f;
+    expected_payload.temperature_2 = 12.0f;
+    expected_payload.temperature_3 = 13.0f;
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: Not an error but you can initialize the structure directly on declaration like this:
```suggestion
    bms_cellboard_cells_temperature_converted_t expected_payload = {
    .cellboard_id = CELLBOARD_ID,
    .offset = 0U,
    .temperature_0 = 10.0f,
    .temperature_1 = 11.0f,
    .temperature_2 = 12.0f,
    .temperature_3 = 13.0f,
};
```

File: test/test_temp/test-temp.c [hunk 13]
```diff
@@ -1,166 +1,283 @@
...
+    TEST_ASSERT_EQUAL_MEMORY_MESSAGE(expected, out, sizeof(out), "temp_dump_values() failed to copy the correct values to the output array");
 }
 
-void test_temp_get_discharge_temp_canlib_payload() {
+void test_temp_get_cells_temp_canlib_payload_values() {
+
+    temp_handler.temperatures[0] = 10.0f;
+    temp_handler.temperatures[1] = 11.0f;
+    temp_handler.temperatures[2] = 12.0f;
+    temp_handler.temperatures[3] = 13.0f;
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: You may consider using a for loop here...

File: test/test_temp/test-temp.c [hunk 14]
```diff
@@ -1,166 +1,283 @@
...
+
+    TEST_ASSERT_EQUAL_MEMORY_MESSAGE(&expected_payload, payload, sizeof(expected_payload), "temp_get_cells_temp_canlib_payload() returned incorrect payload contents");
+}
+
+void test_temp_get_discharge_temp_canlib_payload_values() {
+    temp_handler.discharge_temperatures[0] = 50.0f;
+    temp_handler.discharge_temperatures[1] = 51.0f;
+    temp_handler.discharge_temperatures[2] = 52.0f;
+    temp_handler.discharge_temperatures[3] = 53.0f;
+    temp_handler.discharge_temperatures[4] = 54.0f;
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: Consider using a for loop also here...

File: test/test_temp/test-temp.c [hunk 15]
```diff
@@ -1,166 +1,283 @@
...
 
+    bms_cellboard_discharge_temperature_converted_t expected_payload = { 0 };
+    expected_payload.temperature_0 = 50.0f;
+    expected_payload.temperature_1 = 51.0f;
+    expected_payload.temperature_2 = 52.0f;
+    expected_payload.temperature_3 = 53.0f;
+    expected_payload.temperature_4 = 54.0f;
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]: Same as above, not an error but you may initialize the structure direclty on declaration.

**Review by Tonidotpy — CHANGES_REQUESTED** [review]
- Run `pio check` and solve the warnings of the temp module
- Remember to use LibEagleTRT whenever needed
- Don't overcomplicate things and try to stay coherent

File: Core/Src/bms/temp-api.c [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+
+#ifdef CONF_TEMPERATURE_MODULE_ENABLE
+
+// TODO: Send discharge temperatures
+EAGLETRT_STATIC struct TempHandler temp_handler;
+
+//
+celsius prv_temp_volt_to_celsius(volt_t value) {
+    // Value is converted in V and limited to fit the polynomial range
+    value = EAGLETRT_API_CLAMP(value, TEMP_MIN_LIMIT_V, TEMP_MAX_LIMIT_V); // NOLINT(readability-magic-numbers)
```
Comment (Tonidotpy) [inline Core/Src/bms/temp-api.c]: Instead of disabling the code analyzer in this point replace the macro with constants.
```suggestion
    const volt temp_min_limit = 0.0F;
    const volt temp_max_limit = 3.0F;
    value = EAGLETRT_API_CLAMP(value, temp_min_limit, temp_max_limit);
```

File: Core/Src/bms/temp-api.c [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+           TEMP_COEFF_2 * val2 +
+           TEMP_COEFF_3 * val3 +
+           TEMP_COEFF_4 * val4 +
+           TEMP_COEFF_5 * val5 +
+           TEMP_COEFF_6 * val6;
+}
+
+celsius prv_temp_discharge_volt_to_celsius(volt_t value) {
+    // Value is converted in V and limited to fit the polynomial range
+    value = EAGLETRT_API_CLAMP(value, TEMP_DISCHARGE_MIN_LIMIT_V, TEMP_DISCHARGE_MAX_LIMIT_V); // NOLINT(readability-magic-numbers)
```
Comment (Tonidotpy) [inline Core/Src/bms/temp-api.c]: Same as above.

File: Core/Src/bms/temp-api.c [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+const cells_temp *temp_get_values(void) {
+    return &temp_handler.temperatures;
+}
+
+celsius temp_get_min(void) {
+    celsius min = temp_handler.temperatures[0U];
+    for (size_t i = 0U; i < CELLBOARD_SEGMENT_TEMP_SENSOR_COUNT; ++i) {
+        min = CELLBOARD_MIN(min, temp_handler.temperatures[i]);
```
Comment (Tonidotpy) [inline Core/Src/bms/temp-api.c]:
```suggestion
        min = EAGLETRT_API_MIN(min, temp_handler.temperatures[i]);
```

File: Core/Src/bms/temp-api.c [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+    }
+    return min;
+}
+
+celsius temp_get_max(void) {
+    celsius max = temp_handler.temperatures[0U];
+    for (size_t i = 0U; i < CELLBOARD_SEGMENT_TEMP_SENSOR_COUNT; ++i) {
+        max = CELLBOARD_MAX(max, temp_handler.temperatures[i]);
```
Comment (Tonidotpy) [inline Core/Src/bms/temp-api.c]:
```suggestion
        max = EAGLETRT_API_MAX(max, temp_handler.temperatures[i]);
```

File: platformio.ini [hunk 20]
```diff
@@ -80,7 +82,7 @@ build_src_filter =
 
 check_tool = clangtidy
 check_flags =
-    clangtidy: --checks=-*,cert-*,clang-analyzer-*,bugprone-*,performance-*,readability-*
+    clangtidy: --checks=-*,cert-*,clang-analyzer-*,bugprone-*,performance-*,readability-*,-bugprone-easily-swappable-parameters
```
Comment (Tonidotpy) [inline platformio.ini:84]: Avoid removing all warning for swappable parameters, if necessary remove the warning only for the specific function.
```suggestion
    clangtidy: --checks=-*,cert-*,clang-analyzer-*,bugprone-*,performance-*,readability-*```

File: test/test_temp/test-temp.c [hunk 21]
```diff
@@ -1,166 +1,362 @@
...
+    const enum TempReturnCode rc = temp_init(mock_set_address, NULL);
+
+    TEST_ASSERT_EQUAL_MESSAGE(TEMP_RC_NULL_POINTER, rc, "temp_init() should fail with NULL start_conversion");
+    TEST_ASSERT_EQUAL_MEMORY_MESSAGE(&expected, &temp_handler, sizeof(expected), "temp_init() should not modify temp_handler on failure");
 }
 
-void test_temp_start_conversion_ok() {
-    TEST_ASSERT_EQUAL_MESSAGE(TEMP_OK, temp_start_conversion(), "temp_start_conversion() failed to return TEMP_OK");
-    TEST_ASSERT_TRUE_MESSAGE(htemp.busy, "Module should be busy after starting conversion");
+void test_temp_start_conversion_check_correct_fct_calls() {
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c]:
```suggestion
void test_temp_start_conversion_check_correct_function_calls() {
```

File: test/test_temp/test-temp.c [hunk 22]
```diff
@@ -1,166 +1,362 @@
...
+void test_temp_update_discharge_values_with_valid_values() {
+
+    temp_handler.discharge_temperatures[1] = -999.0f;
+    temp_handler.discharge_temperatures[2] = -999.0f;
+    volt_t values[2] = { 2.0f, 2.1f };
+
+    celsius expected[2] = {
+        prv_temp_discharge_volt_to_celsius(values[0]),
+        prv_temp_discharge_volt_to_celsius(values[1])
+    };
```
Comment (Tonidotpy) [inline test/test_temp/test-temp.c:221]: Don't really like the idea of calling module function in the arrange step. @Bridiro What do you think?

**Review by Tonidotpy — CHANGES_REQUESTED** [review]
Couple of changes. There a couple of doubts that should be cleared.

File: test/test_temp/test-temp.c (hunk 22 above)
Comment (ale-giustina) [reply test/test_temp/test-temp.c:221]: This test is validating the behavior of `temp_update_discharge_values()`, not the correctness of the conversion logic itself. The conversion function already has dedicated unit tests that verify its behavior across midrange and boundary conditions, so it should be treated here as a trusted dependency.

If we replaced these calls with magic numbers, we would either duplicate the conversion logic in the test or rely on constants. Both approaches make the test harder to understand and maintain. This way the test would become prone to fail even though the behavior and logicof `temp_update_discharge_values()` remains correct as any legitimate change to the conversion function would cause this test to fail.

By computing the expected values through `prv_temp_discharge_volt_to_celsius()`, the test stays focused on what it is actually responsible for verifying: the logic of the function, the application of the conversion (and not correctedness) and the correct indexing.

File: Core/Src/bms/temp-api.c [hunk 23]
```diff
@@ -23,8 +23,13 @@ EAGLETRT_STATIC struct TempHandler temp_handler;
 
 //
 celsius prv_temp_volt_to_celsius(volt_t value) {
+
+    // To please clang-tidy, really don't like this
```
Comment (Tonidotpy) [inline Core/Src/bms/temp-api.c:27]: Avoid useless coments in the code they do not help with readability.

**Review by Tonidotpy — APPROVED** [review]
